File: geomesa-avro/src/main/java/AvroSimpleFeature.java
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.vividsolutions.jts.geom.*;
import org.apache.avro.Schema;
import org.apache.avro.SchemaBuilder;
import org.apache.avro.generic.GenericData;
import org.apache.avro.generic.GenericDatumWriter;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.io.BinaryEncoder;
import org.apache.avro.io.DatumWriter;
import org.apache.avro.io.EncoderFactory;
import org.geotools.data.DataUtilities;
import org.geotools.filter.identity.FeatureIdImpl;
import org.geotools.geometry.jts.ReferencedEnvelope;
import org.geotools.util.Converters;
import org.opengis.feature.GeometryAttribute;
import org.opengis.feature.Property;
import org.opengis.feature.simple.SimpleFeature;
import org.opengis.feature.simple.SimpleFeatureType;
import org.opengis.feature.type.AttributeDescriptor;
import org.opengis.feature.type.GeometryDescriptor;
import org.opengis.feature.type.Name;
import org.opengis.filter.identity.FeatureId;
import org.opengis.geometry.BoundingBox;

import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.util.*;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

public class AvroSimpleFeature implements SimpleFeature {

    static final LoadingCache<SimpleFeatureType, Map<String, Class<?>>> typeMapCache = CacheBuilder.newBuilder()
            .maximumSize(100)
            .expireAfterWrite(10, TimeUnit.MINUTES)
            .build(new CacheLoader<SimpleFeatureType, Map<String, Class<?>>>() {
                public Map<String, Class<?>> load(SimpleFeatureType sft) {
                    final Map<String, Class<?>> map = new HashMap<>();
                    for (final AttributeDescriptor attributeDescriptor : sft.getAttributeDescriptors()) {
                        final String name = attributeDescriptor.getLocalName();
                        final Class<?> clazz = attributeDescriptor.getType().getBinding();
                        map.put(name, clazz);
                    }
                    return map;
                }
            });

    static final LoadingCache<SimpleFeatureType, Schema> avroSchemaCache = CacheBuilder.newBuilder()
            .maximumSize(100)
            .expireAfterWrite(10, TimeUnit.MINUTES)
            .build(new CacheLoader<SimpleFeatureType, Schema>() {
                public Schema load(SimpleFeatureType sft) {
                    return generateSchema(sft);
                }
            });

    final Map<String, Class<?>> typeMap;
    final protected SimpleFeatureType sft;
    final String[] names;
    final Object[] values;
    FeatureId id;
    final HashMap<String, Integer> nameIndex = new HashMap<>();
    final HashMap<Object, Object> userData = new HashMap<>();
    final Schema schema;
    protected static final String FEATURE_ID_AVRO_FIELD_NAME = "__fid__";
    protected static final String AVRO_SIMPLE_FEATURE_VERSION = "__version__";
    private static final int VERSION = 1;
    protected static final String AVRO_NAMESPACE = "org.geomesa";

    public AvroSimpleFeature(final FeatureId id, final SimpleFeatureType sft) {
        this.id = id;
        this.values = new Object[sft.getAttributeCount()];
        this.names = new String[sft.getAttributeCount()];
        this.sft = sft;

        int i = 0;
        for (final String name : DataUtilities.attributeNames(sft)) {
            names[i++] = name;
            nameIndex.put(name, sft.indexOf(name));
        }

        try {
            this.schema = avroSchemaCache.get(sft);
            this.typeMap = typeMapCache.get(sft);
        } catch (ExecutionException e) {
            throw new IllegalStateException("Unable to create type map and schema");
        }

    }

    static Schema generateSchema(final SimpleFeatureType sft) {
        SchemaBuilder.FieldAssembler assembler = SchemaBuilder
                .record(sft.getTypeName())
                .namespace(AVRO_NAMESPACE).fields()
                .name(AVRO_SIMPLE_FEATURE_VERSION).type().intType().noDefault()
                .name(FEATURE_ID_AVRO_FIELD_NAME).type().stringType().noDefault();

        for (AttributeDescriptor attributeDescriptor : sft.getAttributeDescriptors()) {
            final String name = attributeDescriptor.getLocalName();
            final Class<?> clazz = attributeDescriptor.getType().getBinding();
            if (clazz == String.class) {
                assembler = assembler.name(name).type().stringType().noDefault();
            } else if (clazz == Integer.class) {
                assembler = assembler.name(name).type().intType().noDefault();
            } else if (clazz == Long.class) {
                assembler = assembler.name(name).type().longType().noDefault();
            } else if (clazz == Double.class) {
                assembler = assembler.name(name).type().doubleType().noDefault();
            } else if (clazz == Float.class) {
                assembler = assembler.name(name).type().floatType().noDefault();
            } else if (clazz == Boolean.class) {
                assembler = assembler.name(name).type().booleanType().noDefault();
            } else if (clazz == UUID.class) {
                assembler = assembler.name(name).type().bytesType().noDefault();
            } else if (clazz == Date.class) {
                // Represent as long (millis)
                assembler = assembler.name(name).type().longType().noDefault();
            } else {
                // Assume string serialization for othe types...
                assembler = assembler.name(name).type().stringType().noDefault();
            }
        }
        return (Schema) assembler.endRecord();
    }

    public void write(OutputStream os) throws IOException {
        final BinaryEncoder encoder = EncoderFactory.get().binaryEncoder(os, null);
        final GenericRecord me = new GenericData.Record(this.schema);
        me.put(AVRO_SIMPLE_FEATURE_VERSION, VERSION);
        me.put(FEATURE_ID_AVRO_FIELD_NAME, this.getID());
        for (int i = 0; i < values.length; i++) {
            final Class<?> clazz = typeMap.get(names[i]);
            if (clazz == String.class || clazz == Integer.class || clazz == Long.class || clazz == Double.class
                    || clazz == Float.class || clazz == Boolean.class) {
                me.put(names[i], values[i]);
            } else if (clazz == UUID.class) {
                final UUID uuid = (UUID) values[i];
                final ByteBuffer bb = ByteBuffer.wrap(new byte[16]);
                bb.putLong(uuid.getMostSignificantBits());
                bb.putLong(uuid.getLeastSignificantBits());
                bb.flip();
                me.put(names[i], bb);
            } else if (clazz == Date.class) {
                final Date d = (Date) values[i];
                final Long l = d.getTime();
                me.put(names[i], l);
            } else if (Geometry.class.isAssignableFrom(clazz)) {
                Geometry geometry = (Geometry) values[i];
                String txt = geometry.toText();
                me.put(names[i], txt);
            } else {
                String txt = Converters.convert(values[i], String.class);
                if (txt == null) { // could not convert?
                    txt = values[i].toString();
                }
                me.put(names[i], txt);
            }
        }
        final DatumWriter<GenericRecord> datumWriter = new GenericDatumWriter<>(this.schema);
        datumWriter.write(me, encoder);
        encoder.flush();
    }

    public SimpleFeatureType getFeatureType() {
        return sft;
    }

    public SimpleFeatureType getType() {
        return sft;
    }

    public void setID(String id) {
        ((FeatureIdImpl) this.id).setID(id);
    }

    public FeatureId getIdentifier() {
        return id;
    }

    public String getID() {
        return id.getID();
    }

    public Object getAttribute(String name) {
        return getAttribute(nameIndex.get(name));
    }

    public Object getAttribute(Name name) {
        return getAttribute(name.getLocalPart());
    }

    public Object getAttribute(int index) throws IndexOutOfBoundsException {
        return this.values[index];
    }

    public void setAttribute(String name, Object value) {
        int i = nameIndex.get(name);
        setAttribute(i, value);
    }

    public void setAttribute(Name name, Object value) {
        setAttribute(name.getLocalPart(), value);
    }

    public void setAttribute(int index, Object value)
            throws IndexOutOfBoundsException {
        values[index] = value;
    }

    public void setAttributes(List<Object> values) {
        for (int i = 0; i < values.size(); i++) {
            setAttribute(i, values.get(i));
        }
    }

    public int getAttributeCount() {
        return values.length;
    }

    public List<Object> getAttributes() {
        return Arrays.asList(this.values);
    }

    public Object getDefaultGeometry() {
        GeometryDescriptor defaultGeometry = sft.getGeometryDescriptor();
        return defaultGeometry != null ? getAttribute(defaultGeometry.getName()) : null;
    }

    public void setAttributes(Object[] object) {
        throw new UnsupportedOperationException();
    }

    public void setDefaultGeometry(Object defaultGeometry) {
        GeometryDescriptor descriptor = sft.getGeometryDescriptor();
        setAttribute(descriptor.getName(), defaultGeometry);
    }

    public BoundingBox getBounds() {
        Object obj = getDefaultGeometry();
        if (obj instanceof Geometry) {
            Geometry geometry = (Geometry) obj;
            return new ReferencedEnvelope(geometry.getEnvelopeInternal(), sft.getCoordinateReferenceSystem());
        }
        return new ReferencedEnvelope(sft.getCoordinateReferenceSystem());
    }

    public GeometryAttribute getDefaultGeometryProperty() {
        throw new UnsupportedOperationException();
    }

    public void setDefaultGeometryProperty(GeometryAttribute defaultGeometry) {
        throw new UnsupportedOperationException();
    }

    public Collection<Property> getProperties() {
        throw new UnsupportedOperationException();
    }

    public Collection<Property> getProperties(Name name) {
        throw new UnsupportedOperationException();
    }

    public Collection<Property> getProperties(String name) {
        throw new UnsupportedOperationException();
    }

    public Property getProperty(Name name) {
        throw new UnsupportedOperationException();
    }

    public Property getProperty(String name) {
        throw new UnsupportedOperationException();
    }

    public Collection<? extends Property> getValue() {
        throw new UnsupportedOperationException();
    }

    public void setValue(Collection<Property> value) {
        throw new UnsupportedOperationException();
    }

    public AttributeDescriptor getDescriptor() {
        throw new UnsupportedOperationException();
    }

    public Name getName() {
        throw new UnsupportedOperationException();
    }

    public Map<Object, Object> getUserData() {
        return userData;
    }

    public boolean isNillable() {
        throw new UnsupportedOperationException();
    }

    public void setValue(Object value) {
        throw new UnsupportedOperationException();
    }

    public void validate() {
    }

    /**
     * override of equals.  Returns if the passed in object is equal to this.
     *
     * @param obj the Object to test for equality.
     * @return <code>true</code> if the object is equal, <code>false</code>
     *         otherwise.
     */
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }

        if (obj == this) {
            return true;
        }

        if (!(obj instanceof AvroSimpleFeature)) {
            return false;
        }

        AvroSimpleFeature feat = (AvroSimpleFeature) obj;

        // this check shouldn't exist, by contract,
        //all features should have an ID.
        if (id == null) {
            if (feat.getIdentifier() != null) {
                return false;
            }
        }

        if (!id.equals(feat.getIdentifier())) {
            return false;
        }

        if (!feat.getFeatureType().equals(sft)) {
            return false;
        }

        for (int i = 0, ii = values.length; i < ii; i++) {
            Object otherAtt = feat.getAttribute(i);

            if (values[i] == null) {
                if (otherAtt != null) {
                    return false;
                }
            } else {
                if (!values[i].equals(otherAtt)) {
                    return false;
                }
            }
        }

        return true;
    }

    /**
     * returns a unique code for this feature
     *
     * @return A unique int
     */
    public int hashCode() {
        return id.hashCode() * sft.hashCode();
    }


}

File: geomesa-avro-scala/src/main/scala/geomesa/avro/scala/AvroSimpleFeature.scala
package geomesa.avro.scala


import collection.JavaConversions._
import com.google.common.cache.{CacheLoader, LoadingCache, CacheBuilder}
import com.vividsolutions.jts.geom.Geometry
import java.io.OutputStream
import java.lang._
import java.nio._
import java.util
import java.util.concurrent.TimeUnit
import org.apache.avro.generic.{GenericDatumWriter, GenericData, GenericRecord}
import org.apache.avro.io.{BinaryEncoder, EncoderFactory}
import org.apache.avro.{SchemaBuilder, Schema}
import org.geotools.data.DataUtilities
import org.geotools.geometry.jts.ReferencedEnvelope
import org.geotools.util.Converters
import org.opengis.feature.{Property, GeometryAttribute}
import org.opengis.feature.simple.{SimpleFeatureType, SimpleFeature}
import org.opengis.filter.identity.FeatureId
import java.util.{Date, UUID, List => JList}
import org.opengis.geometry.BoundingBox
import org.opengis.feature.`type`.Name
import org.opengis.feature.`type`.AttributeDescriptor
import scala.util.Try
import org.apache.avro.SchemaBuilder.BaseFieldTypeBuilder
import org.geotools.feature.GeometryAttributeImpl
import org.geotools.feature.`type`.AttributeDescriptorImpl


class AvroSimpleFeature(id: FeatureId, sft: SimpleFeatureType) extends SimpleFeature {

  import AvroSimpleFeature._

  val values    = Array.ofDim[AnyRef](sft.getAttributeCount)
  val userData  = collection.mutable.HashMap.empty[AnyRef, AnyRef]
  val typeMap   = typeMapCache.get(sft)
  val names     = nameCache.get(sft)
  val nameIndex = nameIndexCache.get(sft)
  val schema    = avroSchemaCache.get(sft)

  def write(datumWriter: GenericDatumWriter[GenericRecord], encoder: BinaryEncoder){
    val record = new GenericData.Record(schema)
    record.put(AvroSimpleFeature.AVRO_SIMPLE_FEATURE_VERSION, AvroSimpleFeature.VERSION)
    record.put(AvroSimpleFeature.FEATURE_ID_AVRO_FIELD_NAME, getID)

    values.zipWithIndex.foreach { case (v, idx) =>
      val x = v == null match {
        case true =>
          v
        case false => typeMap(names(idx)) match {
          case t if primitiveTypes.contains(t) =>
            v

          case t if classOf[UUID].isAssignableFrom(t) =>
            val uuid = v.asInstanceOf[UUID]
            val bb = ByteBuffer.allocate(16)
            bb.putLong(uuid.getMostSignificantBits)
            bb.putLong(uuid.getLeastSignificantBits)
            bb.flip
            bb

          case t if classOf[Date].isAssignableFrom(t) => {
            v.asInstanceOf[Date].getTime
          }

          case t if classOf[Geometry].isAssignableFrom(t) =>
            v.asInstanceOf[Geometry].toText

          case _ =>
            Option(Converters.convert(v, classOf[String])).getOrElse { a: AnyRef => a.toString }
        }
      }
      record.put(names(idx), x)
    }
    datumWriter.write(record, encoder)
    encoder.flush()
  }

  def write(os: OutputStream) {
    val encoder = EncoderFactory.get.binaryEncoder(os, null)
    val datumWriter = new GenericDatumWriter[GenericRecord](schema)
    write(datumWriter, encoder)
  }

  def getFeatureType = sft
  def getType = sft
  def getIdentifier = id
  def getID = id.getID
  def getAttribute(name: String) = nameIndex.get(name).map(getAttribute).getOrElse(throw new NullPointerException("Invalid attribute"))
  def getAttribute(name: Name) = getAttribute(name.getLocalPart)
  def getAttribute(index: Int) = values(index)
  def setAttribute(name: String, value: Object) = setAttribute(nameIndex(name), value)
  def setAttribute(name: Name, value: Object) = setAttribute(name.getLocalPart, value)
  def setAttribute(index: Int, value: Object) = values(index) = value
  def setAttributes(values: JList[Object]) =
    values.zipWithIndex.foreach { case (v, idx) => setAttribute(idx, v) }

  def getAttributeCount = values.length
  def getAttributes: JList[Object] = values.toList
  def getDefaultGeometry: Object = Try(sft.getGeometryDescriptor.getName).map { getAttribute }.getOrElse(null)

  def setAttributes(`object`: Array[Object])= ???

  def setDefaultGeometry(geo: Object) =
    setAttribute(sft.getGeometryDescriptor.getName, geo)

  def getBounds: BoundingBox = getDefaultGeometry match {
    case g: Geometry =>
      new ReferencedEnvelope(g.getEnvelopeInternal, sft.getCoordinateReferenceSystem)
    case _ =>
      new ReferencedEnvelope(sft.getCoordinateReferenceSystem)
  }

  def getDefaultGeometryProperty: GeometryAttribute = {
    val geoDesc = sft.getGeometryDescriptor
    geoDesc != null match {
      case true =>
        new GeometryAttributeImpl(getDefaultGeometry, geoDesc, null)
      case false =>
        null
    }
  }

  def setDefaultGeometryProperty(geoAttr: GeometryAttribute) = geoAttr != null match {
    case true =>
      setDefaultGeometry(geoAttr.getValue)
    case false =>
      setDefaultGeometry(null)
  }

  def getProperties: util.Collection[Property] = ???
  def getProperties(name: Name): util.Collection[Property] = ???
  def getProperties(name: String): util.Collection[Property] = ???
  def getProperty(name: Name): Property = ???
  def getProperty(name: String): Property = ???

  def getValue: util.Collection[_ <: Property] = getProperties

  def setValue(values: util.Collection[Property]) = values.zipWithIndex.foreach { case (p, idx) =>
    this.values(idx) = p.getValue}

  def getDescriptor: AttributeDescriptor = new AttributeDescriptorImpl(sft, sft.getName, 0, Int.MaxValue, true, null)

  def getName: Name = sft.getName

  def getUserData = userData

  def isNillable = true

  def setValue(newValue: Object) = setValue (newValue.asInstanceOf[util.Collection[Property]])

  def validate() = { }
}

object AvroSimpleFeature {

  def apply(sf: SimpleFeature) = {
    val asf = new AvroSimpleFeature(sf.getIdentifier, sf.getFeatureType)
    for (i <- 0 until sf.getAttributeCount) asf.setAttribute(i, sf.getAttribute(i))

    asf
  }

  import scala.collection.JavaConversions._

  val primitiveTypes =
    List(
      classOf[String],
      classOf[Integer],
      classOf[Long],
      classOf[Double],
      classOf[Float],
      classOf[Boolean]
    )

  def loadingCacheBuilder[V <: AnyRef](f: SimpleFeatureType => V) =
    CacheBuilder
      .newBuilder
      .maximumSize(100)
      .expireAfterWrite(10, TimeUnit.MINUTES)
      .build(
        new CacheLoader[SimpleFeatureType, V] {
          def load(sft: SimpleFeatureType): V = f(sft)
        }
      )

  val typeMapCache: LoadingCache[SimpleFeatureType, Map[String, Class[_]]] =
    loadingCacheBuilder { sft =>
      sft.getAttributeDescriptors.map { ad =>
        val name = ad.getLocalName
        val clazz = ad.getType.getBinding
        (name, clazz)
      }.toMap
    }

  val avroSchemaCache: LoadingCache[SimpleFeatureType, Schema] =
    loadingCacheBuilder { sft => generateSchema(sft) }

  val nameCache: LoadingCache[SimpleFeatureType, Array[String]] =
    loadingCacheBuilder { sft => DataUtilities.attributeNames(sft) }

  val nameIndexCache: LoadingCache[SimpleFeatureType, Map[String, Int]] =
    loadingCacheBuilder { sft =>
      DataUtilities.attributeNames(sft).map { name => (name, sft.indexOf(name)) }.toMap
    }

  val datumWriterCache: LoadingCache[SimpleFeatureType, GenericDatumWriter[GenericRecord]] =
    loadingCacheBuilder { sft =>
      new GenericDatumWriter[GenericRecord](avroSchemaCache.get(sft))
    }


  final val FEATURE_ID_AVRO_FIELD_NAME: String = "__fid__"
  final val AVRO_SIMPLE_FEATURE_VERSION: String = "__version__"
  final val VERSION: Int = 1
  final val AVRO_NAMESPACE: String = "org.geomesa"

  def generateSchema(sft: SimpleFeatureType): Schema = {
    val initialAssembler: SchemaBuilder.FieldAssembler[Schema] =
      SchemaBuilder.record(sft.getTypeName)
        .namespace(AVRO_NAMESPACE)
        .fields
        .name(AVRO_SIMPLE_FEATURE_VERSION).`type`.intType.noDefault
        .name(FEATURE_ID_AVRO_FIELD_NAME).`type`.stringType.noDefault

    val result =
      sft.getAttributeDescriptors.foldLeft(initialAssembler) { case (assembler, ad) =>
        val name    = ad.getLocalName
        val binding = ad.getType.getBinding
        val nillable = ad.isNillable
        addField(assembler, name, binding, nillable)
      }

    result.endRecord
  }

  def addField(assembler: SchemaBuilder.FieldAssembler[Schema],
               name: String,
               ct: Class[_],
               nillable: Boolean): SchemaBuilder.FieldAssembler[Schema] = {
    val baseType = if (nillable) assembler.name(name).`type`.nullable() else assembler.name(name).`type`
    ct match {
      case c if classOf[String].isAssignableFrom(c)   => baseType.stringType().noDefault()
      case c if classOf[Integer].isAssignableFrom(c)  => baseType.intType.noDefault
      case c if classOf[Long].isAssignableFrom(c)     => baseType.longType.noDefault
      case c if classOf[Double].isAssignableFrom(c)   => baseType.doubleType.noDefault
      case c if classOf[Float].isAssignableFrom(c)    => baseType.floatType.noDefault
      case c if classOf[Boolean].isAssignableFrom(c)  => baseType.booleanType.noDefault
      case c if classOf[UUID].isAssignableFrom(c)     => baseType.bytesType.noDefault
      case c if classOf[Date].isAssignableFrom(c)     => baseType.longType.noDefault
      case c if classOf[Geometry].isAssignableFrom(c) => baseType.stringType.noDefault
    }
  }

}
